### ISSUE TITLE
Improve required module workflow

### ISSUE THREAD

**Issue by lenaqr** [learning-unlimited/ESP-Website#1827]
Related to and possibly supersedes #335, #1496.

An MIT ESP brought up some concerns about the required module redirection workflow possibly being disorienting to students (and misleading about parents, since the medical form is a required step).

One suggestion, similar to #335:

> Add the checkboxes thingy that appears on student reg appearing on all the modules so that we can get rid of "registration steps" completely. 

Another, possibly orthogonal, idea that I've been thinking about is to have some way for each module to have prerequisites, rather than just each module being required or not; then if any module's prerequisites are not met, it tells you what those prerequisites are and redirects you to them.

A third idea: the whole idea of the registration mainpage in the first place was that you go there first and it gives you a checklist of steps. Maybe running with force_show_required_modules disabled should be a more supported workflow; we could do this by making the mainpage content more dynamic based on which modules the user has yet to complete.

**Comment by mgersh** [comment]
The second point is something I've been thinking about too. A very related thing that would also be useful is the ability to set up workflows where module A always redirects to module B after being filled out, or after the first time it's completed (I really want program creation to work this way), but this is probably harder.

**Comment by imran-techvoyager** [comment]
Hi @willgearty , I’d like to work on this issue and help improve the required module workflow.

After reviewing the redirect logic in base.py (the findModule() flow around required modules), it seems the main usability issue is that when force_show_required_modules is enabled, students are silently redirected to the first incomplete required module before ever seeing the registration main page. However, individual module pages don’t render the checklist/progress UI that exists on studentregcore/mainpage.html, so there’s no orientation context.

As a first step, I propose focusing on the UI consistency layer:

Introduce a shared module.html base template for required modules.

Move the checklist/progress UI into that base template.

Ensure completion state is injected centrally (rather than per-template duplication).

This would address the disorientation concern without immediately altering redirect semantics or data models.

Separately, I’m interested in exploring prerequisites (via a self-referential ManyToMany on ProgramModuleObj) as discussed above, but I think solving template structure and orientation first gives us a safer foundation.

Does starting with the shared module template approach make sense as the next concrete step?

**Comment by willgearty** [comment]
Yeah, a shared template with the checklist sounds like a great idea! @imran-techvoyager

**Comment by willgearty** [comment]
Reopening because there are some other items to be completed here. @imran-techvoyager were you planning on exploring the other things discussed here? Should we make subissues for them?

**Comment by imran-techvoyager** [comment]
@willgearty Thanks for reopening this.

Yes, I’m happy to keep exploring the remaining ideas from the original discussion. It probably makes sense to split them into separate subissues (for example module prerequisites and the `force_show_required_modules` workflow) so they can be tracked independently.

Happy to open subissues for those if that makes sense.

**Comment by willgearty** [comment]
OK, I think the new #4964 and the old #1496 cover the remaining topics from above. I think there are two avenues for this, both of which would require quite a bit of discussion:

1. Hacky solutions that provide a temporary solution. It's not immediately clear to me what that solution would be for either of these items.
2. Including this in the Module Management UI GSoC project (which has admittedly scope crept to "Module Management and Functionality" more broadly) and coming up with real solutions to these as part of the broader redesign of module management and interrelatedness.

**Comment by imran-techvoyager** [comment]
Thanks for the clarification @willgearty , that makes sense.

I agree that both #4964 and #1496 probably fit better as part of the broader module management redesign rather than trying to patch them individually. I’ll treat them as part of the Module Management / Functionality work and incorporate them into the GSoC proposal so we can approach them more holistically.

Happy to continue the discussion there as the proposal evolves.
